### PR TITLE
Add redis

### PR DESCRIPTION
**Pull request by BohdanRysakov**

### CODE REVIEW

File: redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheChatRepository.kt [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+                actualRepository.findChatById(id)
+                    .flatMap { savedChat ->
+                        redisOperations.opsForValue().set("$CHAT_CACHE_KEY_PREFIX$id", savedChat)
+                            .thenReturn(savedChat)
+                            .doOnSuccess {
+                                logger.info("Chat with id {} was saved in cache", savedChat.id)
+                            }
+                    }
+            }
+    }
```
Comment (BohdanRysakov) [inline redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheChatRepository.kt:39]: 5 nesting levels here is bad?

File: redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheMessageRepository.kt [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            }
+    }
+
+    override fun deleteAll(): Mono<Unit> {
+        return actualRepository.deleteAll()
+            .thenMany(
+                redisOperations.scan(ScanOptions.scanOptions().match("$MESSAGE_CACHE_KEY_PREFIX*").build())
+            )
+            .flatMap {
+                redisOperations.opsForValue().delete(it)
```
Comment (BohdanRysakov) [inline redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheMessageRepository.kt:50]: Am I right that Redis is pretty fast so we can perform 1 by 1 deletion? 
(I mean in mongo it was bad practice - creating N delete requests)

File: redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheChatRepository.kt (hunk 1 above)
Comment (ajax-shulepa-a) [reply redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheChatRepository.kt:38]: 5 nesting levels always bad

File: redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheMessageRepository.kt (hunk 2 above)
Comment (ajax-shulepa-a) [reply redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheMessageRepository.kt:50]: redis is fast, but think how many network time you'll use

File: core/src/main/kotlin/rys/ajaxpetproject/repository/impl/ChatRepositoryImpl.kt [hunk 3]
```diff
@@ -131,24 +142,27 @@ class ChatRepositoryImpl(
...
             .flatMap { messageIds ->
-                messageRepository.deleteMessagesByIds(messageIds)
+                messageRepository.deleteMessagesByIds(messageIds).thenReturn(messageIds)
             }
+            .flatMap { messageIds ->
+                removeMessages(messageIds, chatId)
+            }
+
+
     }
```
Comment (ajax-shulepa-a) [inline core/src/main/kotlin/rys/ajaxpetproject/repository/impl/ChatRepositoryImpl.kt:167]:
```suggestion
            }
    }
```

File: redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheChatRepository.kt [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+    override fun deleteAll(): Mono<Unit> {
+        return actualRepository.deleteAll()
+            .thenMany(
+                redisOperations.scan(ScanOptions.scanOptions().match("$CHAT_CACHE_KEY_PREFIX*").build())
+            )
+            .flatMap {
+                redisOperations.opsForValue().delete(it)
+            }
+            .then(Unit.toMono())
```
Comment (ajax-shulepa-a) [inline redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheChatRepository.kt:58]:
```suggestion
            .thenReturn(Unit)
```

File: redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheChatRepository.kt [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+                    .flatMap { updatedChat ->
+                        redisOperations.opsForValue().set("$CHAT_CACHE_KEY_PREFIX${updatedChat.id!!}", updatedChat)
+                            .thenReturn(Unit)
+                    }.doOnSuccess {
+                        logger.info(
+                            "User with id {} was removed from chat with id {} in cache",
+                            userId,
+                            chatId
+                        )
+                    }
```
Comment (ajax-shulepa-a) [inline redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheChatRepository.kt:94]: refactor it to make function shorter. applicable to all similar places

**Review by ajax-shulepa-a — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/rys/ajaxpetproject/repository/impl/ChatRepositoryImpl.kt (hunk 3 above)
Comment (BohdanRysakov) [reply core/src/main/kotlin/rys/ajaxpetproject/repository/impl/ChatRepositoryImpl.kt:167]: Done

File: redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheChatRepository.kt (hunk 5 above)
Comment (BohdanRysakov) [reply redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheChatRepository.kt]: Done

File: redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheChatRepository.kt (hunk 1 above)
Comment (BohdanRysakov) [reply redis/src/main/kotlin/rys/ajaxpetproject/redis/repository/CacheChatRepository.kt]: solved

**Review by ajax-shulepa-a — APPROVED** [review]

**Pull request merged**